src/Futhark/LSP/Tool.hs

```diff
@@ -0,0 +1,37 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Futhark.LSP.Tool (getHoverInfoFromState) where
+
+import qualified Data.Text as T
+import Futhark.Compiler (lpImports)
+import Futhark.LSP.State (State (..))
+import Futhark.Util.Loc (srclocOf)
+import Futhark.Util.Pretty (pretty)
+import Language.Futhark.Query
+  ( AtPos (AtName),
+    BoundTo (BoundModule, BoundModuleType, BoundTerm, BoundType),
+    Pos (Pos),
+    atPos,
+  )
+import Language.Futhark.Syntax (locStr)
+
+getHoverInfoFromState :: State -> Maybe FilePath -> Int -> Int -> IO (Maybe T.Text)
+getHoverInfoFromState state (Just path) l c = do
+  case stateProgram state of
+    Nothing -> pure Nothing
+    Just loaded_prog -> do
+      let imports = lpImports loaded_prog
+      case atPos imports $ Pos path l c 0 of
+        Nothing -> pure Nothing
+        Just (AtName qn def _loc) -> do
+          case def of
+            Nothing -> pure Nothing
+            Just (BoundTerm t defloc) -> do
+              pure $ Just $ T.pack $ pretty qn ++ " :: " ++ pretty t ++ "\n\n" ++ "**Definition: " ++ locStr (srclocOf defloc) ++ "**"
```

Comment:
Why two colons?

Reply:
I was thinking about copying the haskell-language-server style here. Say if we hover on `qn` here, HLS will show `qn :: QualName VName`

Reply:
That is because Haskell uses `::` for type ascription.  Futhark (and most other functional languages) use `:`.

Reply:
Good point, fixed in https://github.com/diku-dk/futhark/pull/1618/commits/22a5de28199481672a7eddae94e7765f2423889e